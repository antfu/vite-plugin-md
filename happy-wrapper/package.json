{
  "name": "happy-wrapper",
  "version": "0.1.0",
  "description": "A wrapper around the HappyDOM which makes it more functional in signature and provides some helpful utilities beyond the DOM implementation",
  "license": "ISC",
  "author": "Ken Snyder <ken@ken.net>",
  "main": "dist/index.js",
  "module": "dist/main.mjs",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsup src/index.ts --format cjs,esm --dts --sourcemap external",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix",
    "test": "vitest",
    "test:ui": "vitest --ui",
    "watch": "tsup src/index.ts --format cjs,esm --dts --sourcemap external --watch"
  },
  "dependencies": {
    "callsites": "^3.1.0",
    "fp-ts": "^2.12.1",
    "happy-dom": "^3.2.0",
    "tsup": "^5.12.7"
  },
  "devDependencies": {
    "@vitest/ui": "^0.12.4",
    "vitest": "^0.12.4"
  }
}
