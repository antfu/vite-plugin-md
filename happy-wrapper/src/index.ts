export { Document, DocumentFragment as Fragment, IElement, IText } from 'happy-dom'

export * from './attributes'
export * from './create'
export * from './diagnostics'
export * from './errors'
export * from './happy-types'
export * from './nodes'
export * from './select'
export * from './type-guards'
export * from './utils'
