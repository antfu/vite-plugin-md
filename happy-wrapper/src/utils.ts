import { flow, pipe } from 'fp-ts/lib/function'
import { createDocument, createElement, createFragment, createTextNode } from './create'
import { HappyMishap } from './errors'
import type { Container, ContainerOrHtml, HTML, NodeSolver, NodeSolverReceiver, NodeType } from './happy-types'
import { isDocument, isElement, isElementLike, isFragment, isTextNode } from './type-guards'

/**
 * Determines the "content-type" of a given node
 */
export const getNodeType = (node: Container | HTML): NodeType => {
  if (typeof node === 'string')
    return 'html'

  return isTextNode(node)
    ? 'text'
    : isElement(node)
      ? 'element'
      : isDocument(node)
        ? 'document'
        : isFragment(node)
          ? 'fragment'
          : 'node'
}

/**
 * A helper utility to help convert DOM nodes or HTML to common type.
 *
 * Start by providing the _exclusions_ you want to make for input. By default, all
 * `Container` types are allowed along with `HTML`
 */
export const solveForNodeType: NodeSolver = (_ = undefined as never) => {
  const solver = <EE extends NodeType, OO>(): NodeSolverReceiver<EE, OO> => ({
    solver: s =>
      (node, parent) => {
        if (node === null)
          throw new Error('Value passed into solver was NULL!')
        if (node === undefined)
          throw new Error('Value passed into solver was UNDEFINED!')

        const type = getNodeType(node)
        if (type in s) {
          const fn = (s as any)[type]
          return fn(node, parent)
        }
        else {
          if (type === 'node' && 'element' in s && isElement(node)) {
            const fn = (s as any).element
            return fn(node, parent)
          }
          else if (type === 'node' && 'text' in s && isTextNode(node)) {
            const fn = (s as any).text
            return fn(node)
          }

          throw new HappyMishap(`Problem finding "${type}" in solver.`, { name: `solveForNodeType(${type})` })
        }
      }
    ,
  })
  return {
    outputType: () => solver(),
    mirror: () => solver(),
  }
}

/**
 * Ensures any Container, array of Containers, or even HTML or HTML[] are all
 * normalized down to just HTML.
 */
export function toHtml<D extends ContainerOrHtml | ContainerOrHtml[] | null>(node: D): HTML {
  if (node === null)
    return ''

  const n = (!Array.isArray(node) ? [node] : node) as ContainerOrHtml[]
  try {
    const results = n.map((i) => {
      const convert = solveForNodeType()
        .outputType<HTML>()
        .solver({
          html: h => h,
          text: t => t.textContent,
          element: e => e.outerHTML,
          node: (ne) => {
            if (isElement(ne))
              convert(ne)
            if (isTextNode(ne))
              convert(ne)

            throw new Error(
              `Unknown node type detected while converting to HTML: [ name: ${ne.nodeName}, type: ${ne.nodeType}, value: ${ne.nodeValue} ]`,
            )
          },
          document: d => `<html>${d.head.hasChildNodes() ? d.head.outerHTML : ''}${d.body.outerHTML}</html>`,
          fragment: (f) => {
            if (isElementLike(f))
              return f.firstElementChild.outerHTML

            else
              return f.childNodes.map(c => convert(c, f)).join('')
          },
        })

      return convert(i)
    })

    return results.join('')
  }
  catch (e) {
    if (Array.isArray(node))
      throw new HappyMishap(`Problem converting an array of ${n.length} nodes [${n.map(i => getNodeType(i as any)).join(', ')}] to HTML`, { name: 'toHTML([...])', inspect: ['first node', node[0]], error: e })

    else
      throw new HappyMishap(`Problem converting "${getNodeType(node)}" to HTML!`, { name: 'toHTML(getNodeType(node))', inspect: node, error: e })
  }
}

/**
 * Clones most DOM types
 */
export function clone<T extends Container | HTML>(container: T): T {
  const clone = solveForNodeType()
    .mirror()
    .solver({
      html: h => `${h}`,
      fragment: flow(toHtml, createFragment),
      document: (d) => {
        return createDocument(d.body.innerHTML, d.head.innerHTML)
      },
      element: e => pipe(e, toHtml, createElement),
      node: flow(toHtml, createFragment, f => f.firstElementChild ? f.firstElementChild : f.firstChild),
      text: flow(toHtml, createTextNode),
    })

  return clone(container)
}

/**
 * ensures that a given string doesn't have any HTML inside of it
 */
export function safeString(str: string): string {
  const node = createFragment(str)
  return node.textContent
}
