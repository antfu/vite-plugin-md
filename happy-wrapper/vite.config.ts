/// <reference types="vitest" />
import { defineConfig } from 'vite'

// used for testing, library code uses TSUP to build exports
export default defineConfig(() => ({
  test: {
    dir: 'test',
  },
  plugins: [

  ],
}))
