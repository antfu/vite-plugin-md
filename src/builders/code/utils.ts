
import type { CodeFenceMeta, CodeOptions, LineCallback, PrismOptions, ShikiOptions } from './types'
import { Highlighter } from './types'

export function isValidLanguage(h: Highlighter, _meta: CodeFenceMeta): boolean {
  if (h === Highlighter.prism) {
    //
  }
  else {
    //
  }
}

/** Type Guard to detect options hash for Prism */
export function usesPrismHighlighting(h: CodeOptions): h is PrismOptions {
  return h.engine === Highlighter.prism
}

/** Type Guard to detect options hash for Shiki */
export function usesShikiHighlighting(h: CodeOptions): h is ShikiOptions {
  return h.engine === Highlighter.shiki
}

export function isLineCallback(cb?: false | string | LineCallback): cb is LineCallback {
  return !!(cb && typeof cb === 'function')
}
