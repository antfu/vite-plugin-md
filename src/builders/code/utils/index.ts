export * from './prism-languages'
export * from './highlighting'
export * from './trace'
export * from './log'
