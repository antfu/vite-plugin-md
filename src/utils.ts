export function getVueVersion(defaultVersion = '3.2.0') {
  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    let v = require('vue')
    if (v.default)
      v = v.default
    return v.version || defaultVersion
  }
  catch (e) {
    return defaultVersion
  }
}
