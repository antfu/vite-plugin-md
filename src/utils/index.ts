export * from './core'
export * from './fp'
export * from './createFnWithProps'
export * from './valueOrCallback'
