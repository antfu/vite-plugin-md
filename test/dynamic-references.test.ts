import { describe, expect, it } from 'vitest'
import { getAttribute, select } from '@yankeeinlondon/happy-wrapper'
import { composeFixture } from './utils'

const alt = getAttribute('alt')
const src = getAttribute('src')
const href = getAttribute('href')

describe('loading images and links from frontmatter props', () => {
  it('locally referenced image', async () => {
    const { html } = await composeFixture('get-the-picture')
    const sel = select(html).findFirst('img', 'did not find any images!')
    expect(alt(sel)).toBe('local cat')
    expect(src(sel)).toBe('{{localCat}}')
  })

  it('externally referenced image', async () => {
    const { html } = await composeFixture('get-the-picture')
    const sel = select(html).findAll('img')[1]
    expect(alt(sel)).toBe('remote cat')
    expect(src(sel)).toBe('{{remoteCat}}')
  })

  it('frontmatter based link has retains curly brackets', async () => {
    const { html } = await composeFixture('get-the-picture')
    const sel = select(html).findFirst('a', 'found no links!')
    expect(href(sel)).toBe('{{catLink}}')
  })

  it('snapshot test', async () => {
    const { html } = await composeFixture('get-the-picture')

    expect(html).toMatchSnapshot()
  })
})
