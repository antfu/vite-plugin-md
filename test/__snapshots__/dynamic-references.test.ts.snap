// Vitest Snapshot v1

exports[`loading images and links from frontmatter props > snapshot test 1`] = `
"<div class=\\"markdown-body\\"><h2>A Local Cat</h2>
<p><img src=\\"{{localCat}}\\" alt=\\"local cat\\"></p>
<h2>A Remote Cat</h2>
<p><img src=\\"{{remoteCat}}\\" alt=\\"remote cat\\"></p>
<p>for more on cats, check out this link: <a href=\\"{{catLink}}\\">link</a></p>
<p>{{catLink}}</p>
</div><style type=\\"text/css\\">
:root {
    --md-text-color: #111827;
    --md-code-background: rgba(27,31,35,.05);
    --code-font: ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, \\"Liberation Mono\\", \\"Courier New\\", monospace;
  }
html.dark {
    --md-text-color: #e5e7eb;
    --md-code-background: rgba(229, 231, 235, 0.65);
  }


  .markdown-body {
    color: var(--md-text-color);
  }


  .markdown-body code {
    font-size: 85%;
    padding: 0.2em 0.3em;
    margin-left: 0.1rem;
    margin-right: 0.1rem;
    background-color: var(--md-code-background);
    border-radius: 3px;
    font-family: var(--code-font);
    color: #333;
    white-space: pre;
  }


  .markdown-body blockquote {
    margin-top: 0;
    margin-bottom: 0.25rem;
    padding: 0 1em;
    color: #6a73737d;
    border-left: 0.25rem solid #dfe2e5;
  }


  .markdown-body a {
    cursor: pointer;
  }


  .markdown-body a:hover {
    background-color: rgba(0,0,0,.5);
    font-weight: 700;
    border-radius: 0.2rem 0.4rem;
  }
</style>"
`;
